packages/utils/src/document/document-query-progress/validate-doc-query-progress.ts

```diff
@@ -0,0 +1,255 @@
+ *
+ * This will:
+ *    - create a new facility
+ *    - create a new patient for each patient in `patientsToCreate`
+ *    - check if the patient is linked
+ *    - start a doc query for each patient
+ *    - check the doc query progress for each patient
+ *    - log the result of all the doc queries
+ *    - delete the patient after the doc query is complete
+ *
+ *
+ * Update the respective env variables and run `ts-node validate-doc-query-progress.ts`
+ *
+ * Note: You can use the populate-patients.ts script to populate
+ * the patients with documents you wish to query here which should be the same as
+ * the ones seen in ./patients.
+ */
+
+async function main() {
+  const facility = await metriportApi.createFacility(createFacility);
+
+  const createdPatients: PatientDTO[] = [];
+
+  for (const patientToCreate of patientsToCreate) {
+    const patient = await metriportApi.createPatient(patientToCreate, facility.id);
+
+    console.log("Patient created:", patient.id, patient.firstName, patient.lastName);
+
+    createdPatients.push(patient);
+  }
+
+  await validateDocQueryProgress(createdPatients, facility.id);
+
+  if (resetAndRunAgain) {
+    await validateDocQueryProgress(createdPatients, facility.id, true);
+  }
+
+  for (const patient of createdPatients) {
+    await internalApi.delete(
+      `/internal/patient/${patient.id}?facilityId=${facility.id}&cxId=${cxId}`
+    );
+  }
+}
+
+const validateDocQueryProgress = async (
+  createdPatients: PatientDTO[],
+  facilityId: string,
+  resetAndRunAgain = false
+): Promise<void> => {
+  const promises = createdPatients.map(async patient => {
+    const isPatientLinked = await checkPatientLinkingStatus(patient, facilityId);
+
+    console.log(`isPatientLinked: ${patient.firstName} ${patient.lastName}`, isPatientLinked);
+
+    if (isPatientLinked) {
+      return await queryPatientDocs(patient, facilityId, resetAndRunAgain);
+    }
+
+    return { patient, successful: false };
+  });
+
+  const resultPromises = await Promise.allSettled(promises);
+
+  const fulfilled = resultPromises.flatMap(p => (p.status === "fulfilled" ? p.value : []));
+
+  console.log("Patient statuses:", JSON.stringify(fulfilled, null, 2));
+
+  const successfulPatientQueries = fulfilled.filter(p => p.successful);
+
+  const overallSuccess = successfulPatientQueries.length === createdPatients.length;
+
+  console.log(
+    `The doc query was completed for all patients${
+      resetAndRunAgain ? " (resetAndRunAgain)" : ""
+    } - status were all queries successful:`,
+    overallSuccess
+  );
+};
+
+async function checkPatientLinkingStatus(
+  patient: PatientDTO,
+  facilityId: string
+): Promise<boolean> {
+  let isPatientLinked = await checkPatientLinked(patient, facilityId);
+
+  while (!isPatientLinked) {
+    await sleep(CREATE_PATIENT_SLEEP.asMilliseconds());
+
+    const retryIsPatientLinked = await executeWithRetries(
+      () => checkPatientLinked(patient, facilityId),
+      3,
+      500
+    );
```

Comment:
Because it fails to link i dont want it to run infinitely

Reply:
Oh, I see, if it fails it will return `undefined` and then we set to false and break the while.

TBH, it would be much simpler to just have a try/catch inside the while and not use `executeWithRetries`... :)

Reply:
You're saying instead of `executeWithRetries` to have a try catch logic that will have some sort of retry logic to make sure it runs a few times before failing?

Reply:
I'm assuming we should expect some failures while executing this? Like, getting a 4xx/5xx from the API?

NABD, though, was just wondering if we could simplify the code. TBH, the `executeWithRetries` should throw at the end, the fact it returns `undefined` is confusing. 😅